### PR TITLE
HBW-143 Dynamic delete_if and disable_if options for select

### PR DESCRIPTION
**Pull request by vladdoncheg**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,9 +11,11 @@ See [#296](https://github.com/latera/homs/pull/296) for more details.
 - [#497](https://github.com/latera/homs/pull/497) Handle connection error when try to print order.
 - [#499](https://github.com/latera/homs/pull/499) Increase memory limit for worker.
 - [#289](https://github.com/latera/homs/pull/289) Make conditions `disable_if` and `delete_if` dynamic.
+- [#289](https://github.com/latera/homs/pull/512) Add dynamic `disable_if` and `delete_if` options for select.
 
 ### Bugfixes
 - [#298](https://github.com/latera/homs/pull/298) Keep BP variables values after submit if they are not on form or disabled.
+- [#289](https://github.com/latera/homs/pull/512) Fix setting not-nullable select' value with blank variable
```
Comment (sugarfree1) [inline CHANGELOG.md:18]: image

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 2]
```diff
@@ -216,9 +216,38 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
       }, {});
 
-      this.setState({ formValues: variables });
+      this.setState({ formValues: this.updateSelectVariables(variables) });
     };
 
+    updateSelectVariables = (variables) => {
+      const flattenFields = this.flatFields(this.props.form.fields);
+
+      const fieldNamesforUpdate = flattenFields.filter(
```
Comment (sugarfree1) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:225]: i think you need separate function here

and please use `field.choices.length > 0`

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 3]
```diff
@@ -216,9 +216,38 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
+        field => field.type === 'select' && field.mode === 'select' && field.nullable === false && field.choices.length
+      ).map(field => field.name);
+
+      fieldNamesforUpdate.forEach(((fieldName) => {
+        const updatingField = flattenFields.find(field => field.name === fieldName);
+        let value = updatingField.choices[0];
+
+        if (Array.isArray(value)) {
+          [value] = value;
```
Comment (sugarfree1) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:234]: as for me, then it's better to use [0]

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 4]
```diff
@@ -216,9 +216,38 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
+        }
+
+        variables[fieldName] = value;
+      }));
+
+      return variables;
+    }
+
+    flatFields = fields => fields.flatMap((field) => {
```
Comment (sugarfree1) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:243]: looks like you need separate func here as well

**Review by sugarfree1 — CHANGES_REQUESTED** [review]

File: hbw/app/javascript/packs/hbw/components/form.js.jsx (hunk 3 above)
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/form.js.jsx:234]: I tried, but linter shows `Use array destructuring (eslintprefer-destructuring)`

File: hbw/app/javascript/packs/hbw/components/form.js.jsx (hunk 4 above)
Comment (vladdoncheg) [reply hbw/app/javascript/packs/hbw/components/form.js.jsx:241]: For what? This function performs one action - return flatten fields from multidimensional array.

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 5]
```diff
@@ -216,9 +216,44 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
+      const flattenFields = this.getFlattenFields(this.props.form.fields);
+
+      const fieldNamesforUpdate = this.getFilteredFieldNames(flattenFields);
+
+      fieldNamesforUpdate.forEach(((fieldName) => {
+        const updatingField = flattenFields.find(field => field.name === fieldName);
+        let value = updatingField.choices[0];
+
+        if (Array.isArray(value)) {
+          [value] = value;
```
Comment (GideonEXE) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:232]: How about changing linter rules?

Comment (GideonEXE) [reply hbw/app/javascript/packs/hbw/components/form.js.jsx:232]: Or you can rename variable

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 6]
```diff
@@ -216,9 +216,44 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
+
+      fieldNamesforUpdate.forEach(((fieldName) => {
+        const updatingField = flattenFields.find(field => field.name === fieldName);
+        let value = updatingField.choices[0];
+
+        if (Array.isArray(value)) {
+          value = value[0];
+        }
+
+        variables[fieldName] = value;
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:235]: Are you sure you need to modify the input array?

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 7]
```diff
@@ -216,9 +216,44 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
+
+        if (Array.isArray(value)) {
+          value = value[0];
+        }
+
+        updatingVariables[fieldName] = value;
+      }));
+
+      return updatingVariables;
+    }
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:239]: I think it's better to do this inside general `reduce` not after that.

Also, let's remove the variable `value` mutation.

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 8]
```diff
@@ -216,9 +216,44 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
+
+    getFlattenFields = fields => fields.flatMap((field) => {
+      if (field.type === 'group') {
+        return field.fields;
+      } else {
+        return field;
+      }
+    });
+
+    getFilteredFieldNames = (fields) => {
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:249]: You are able to get names of select fields with its default values at once.

File: hbw/app/javascript/packs/hbw/components/form.js.jsx [hunk 9]
```diff
@@ -216,9 +216,44 @@ modulejs.define('HBWForm', ['React', 'jQuery', 'HBWError', 'HBWFormDatetime',
...
+      const filteredFields = fields.filter(field => (
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/form.js.jsx:250]: Unnecessary interim variable.

**Review by TimAle — CHANGES_REQUESTED** [review]

**Review by TimAle — APPROVED** [review]
Fine by me

**Review by sugarfree1 — APPROVED** [review]

**Pull request merged**